### ISSUE TITLE
whenever outside react/supply is no longer allowed (one broken module + …?)

### ISSUE THREAD

**Issue by AlexDaniel** [rakudo/rakudo#1428]
This change was done intentionally with the idea that nobody probably used it anyway. Turns out that there is one module [that](https://github.com/ufobat/HTTP-Server-Ogre/blob/f96382986ebc8704654baf034cd8c25f597f28a7/lib/HTTP/Server/Ogre/Http2Protocol.pm6#L30) [did](https://github.com/ufobat/HTTP-Server-Ogre/blob/245824329d2abe90e10ebd99a0512b533dd12c18/lib/HTTP/Server/Ogre.pm6#L92-L106), so this raised a question whether this change is justified within v6.c (even though it was not specced).

Basically, this is no longer allowed:
```code
sub foo { whenever Promise.in(2) { say ‘hello’ } }; react foo
```

Result (2017.12):
```
hello
```

Result (HEAD(5cbd55281))
```
===SORRY!=== Error while compiling /tmp/d2Eq46gIDb
Cannot have a 'whenever' block outside the scope of a 'supply' or 'react' block
at /tmp/d2Eq46gIDb:1
------> sub foo { whenever⏏ Promise.in(2) }; react foo
```

See jnthn's opinion here: https://irclog.perlgeek.de/perl6-dev/2018-01-21#i_15717935

**Comment by AlexDaniel** [comment]
FWIW I have no opinion about this at the moment, comments are welcome.

**Comment by AlexDaniel** [comment]
Ah, also, something to consider. This DWIMs:
```
sub foo {
    default { say ‘default’ }
};
given 42 {
    when 30 { say ‘thirty ’ };
    foo;
}
```

I remember the ability to dislocate given/when/default like that being advertised as a feature. Personally I don't find dislocated whenever's as weird because of expected strange consistency. That alone probably does not make a good argument, but it's just some food for thought.

**Comment by jnthn** [comment]
> This change was done intentionally with the idea that nobody probably used it anyway. 

Well, more like "hoping nobody used it, because it was never intended to work". I designed the `react/`supply`/whenever` syntax, and from the start realized that without enforcing lexical nesting of `whenever` then both the compiler and, as importantly, the programmer would have a harder time reasoning about the stream processing taking place. Unfortunately, I failed to put in enforcement of that at the time (but somehow believed I had, and was surprised a couple of weeks ago that it was missing).

Anyway, the question isn't whether or not `whenever` outside of `supply`/`react` should be supported; that it worked is an unfortunate implementation accident, and allowing it blocks not only the long-planned single-whenever handling that allows elimination of concurrency control, but also would stand in the way of future ideas for this construct.

What we should decide now is whether or not this enforcement, and thus the things that rely upon it, should be made a 6.d feature (and thus require a `use v6.d.PREVIEW`). That way, any affected code can stick in a `use v6.c` and continue to work (though a factor of 3 performance difference in message throughput will probably be a strong incentive to not need that).

**Comment by lizmat** [comment]
+1 for keeping it 6.c with potentially breaking code.  The 3x performance difference should indeed be a good incentive to fix the code.

**Comment by jnthn** [comment]
Alright. I don't have time/energy to look at it today, alas. Should be able to squeeze it in tomorrow.

**Comment by jnthn** [comment]
Alternatively, if anyone wants to do it, it's not actually that hard provided the language version is being readily stashed away (say, in a contextual var) to conditionalize the assertion and the optimizations.

**Comment by zoffixznet** [comment]
> if anyone wants to do it

Sorry, do what? lizmat++ voted to keep it as is (I agree, as long as 6.c spec did not have to be modified to accomodate).

I can do the work, but seems like leaving things as is is what people are calling for.

**Comment by zoffixznet** [comment]
@AlexDaniel [response for the above](https://irclog.perlgeek.de/perl6-dev/2018-01-21#i_15719469). I guess this blocker can be closed then?

**Comment by jnthn** [comment]
So in summary:

* I misunderstood what lizmat wrote, reading it as the opposite as what it says. D'oh.
* I remain willing to try and make it 6.d.PREVIEW-only tomorrow, if needed.
* I delegate the final call to @AlexDaniel

**Comment by AlexDaniel** [comment]
Unfortunately, I'm on the other side on this one. I remember reading a few times on reddit and elsewhere people wondering if perl 6 is stable enough, and others pointing out that it is very stable since 2015.12 release. Let's not give reasons to doubt that by hot patching stuff within v6.c, especially when the user is clearly not at fault by thinking that the construct is valid. We were able to get away with it in the past, but as the time goes on this should stop being the norm.

The performance boost is great of course, and maybe that gives people more incentive to try out v6.d.PREVIEW and us to put more effort into v6.d release to make it happen earlier.

@jnthn please try it, but let me know if it turns out to be impossible or too hard and we'll adjust accordingly.

Also, I created a ticket for that module so that they are aware of the upcoming v6.d change. https://github.com/ufobat/HTTP-Server-Ogre/issues/3

**Comment by AlexDaniel** [comment]
Just for the record, there was a discussion about this today.

https://irclog.perlgeek.de/perl6/2018-02-06#i_15781453
https://irclog.perlgeek.de/perl6-dev/2018-02-06#i_15783322
